In [85]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [86]:
df = pd.read_csv("titanic.csv")

df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [87]:
df.shape

(891, 12)

**There are 891 sample for us to analyse**

**Because PassengerId, Name and Ticket cannot convert to real number and has minor impact on the result, then I remove them**

In [88]:
df.drop(['PassengerId','Name', 'Ticket'], axis=1, inplace=True)

df.head(10)

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    male  22.0      1      0   7.2500   NaN        S
1         1       1  female  38.0      1      0  71.2833   C85        C
2         1       3  female  26.0      0      0   7.9250   NaN        S
3         1       1  female  35.0      1      0  53.1000  C123        S
4         0       3    male  35.0      0      0   8.0500   NaN        S
5         0       3    male   NaN      0      0   8.4583   NaN        Q
6         0       1    male  54.0      0      0  51.8625   E46        S
7         0       3    male   2.0      3      1  21.0750   NaN        S
8         1       3  female  27.0      0      2  11.1333   NaN        S
9         1       2  female  14.0      1      0  30.0708   NaN        C

**Now I convert Sex into binary number**

In [89]:
from sklearn.preprocessing import LabelEncoder

In [90]:
LE = LabelEncoder()

In [91]:
df['Sex'] = LE.fit_transform(df['Sex'])

df.head(10)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    1  22.0      1      0   7.2500   NaN        S
1         1       1    0  38.0      1      0  71.2833   C85        C
2         1       3    0  26.0      0      0   7.9250   NaN        S
3         1       1    0  35.0      1      0  53.1000  C123        S
4         0       3    1  35.0      0      0   8.0500   NaN        S
5         0       3    1   NaN      0      0   8.4583   NaN        Q
6         0       1    1  54.0      0      0  51.8625   E46        S
7         0       3    1   2.0      3      1  21.0750   NaN        S
8         1       3    0  27.0      0      2  11.1333   NaN        S
9         1       2    0  14.0      1      0  30.0708   NaN        C

In [92]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int32  
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int32(1), int64(4), object(2)
memory usage: 59.3+ KB


In [93]:
df.describe()

Survived      Pclass         Sex         Age       SibSp       Parch  \
count  891.000000  891.000000  891.000000  714.000000  891.000000  891.000000   
mean     0.383838    2.308642    0.647587   29.699118    0.523008    0.381594   
std      0.486592    0.836071    0.477990   14.526497    1.102743    0.806057   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    2.000000    0.000000   20.125000    0.000000    0.000000   
50%      0.000000    3.000000    1.000000   28.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   38.000000    1.000000    0.000000   
max      1.000000    3.000000    1.000000   80.000000    8.000000    6.000000   

             Fare  
count  891.000000  
mean    32.204208  
std     49.693429  
min      0.000000  
25%      7.910400  
50%     14.454200  
75%     31.000000  
max    512.329200

In [94]:
df['Cabin'].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

**There are many type of cabin and with my basic competence, I will drop this column**

In [95]:
df.drop('Cabin', axis=1, inplace=True)

df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    1  22.0      1      0   7.2500        S
1         1       1    0  38.0      1      0  71.2833        C
2         1       3    0  26.0      0      0   7.9250        S
3         1       1    0  35.0      1      0  53.1000        S
4         0       3    1  35.0      0      0   8.0500        S

**Embarked column has 2 missing values. Then, I fill the empty entries with the most frequency value in Embarked**

In [96]:
df['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [97]:
df['Embarked'].fillna('S', inplace=True)

df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    1  22.0      1      0   7.2500        S
1         1       1    0  38.0      1      0  71.2833        C
2         1       3    0  26.0      0      0   7.9250        S
3         1       1    0  35.0      1      0  53.1000        S
4         0       3    1  35.0      0      0   8.0500        S

**I convert Embraked columns into 3 other columns by using get_dummies function()**

In [98]:
dummy = pd.get_dummies(df['Embarked'])

dummy.head(10)

C  Q  S
0  0  0  1
1  1  0  0
2  0  0  1
3  0  0  1
4  0  0  1
5  0  1  0
6  0  0  1
7  0  0  1
8  0  0  1
9  1  0  0

**And I drop the last column**

In [99]:
dummy.drop('S', axis=1, inplace=True)

dummy

C  Q
0    0  0
1    1  0
2    0  0
3    0  0
4    0  0
..  .. ..
886  0  0
887  0  0
888  0  0
889  1  0
890  0  1

[891 rows x 2 columns]

In [100]:
new_df = pd.concat([df, dummy], axis=1)

new_df.head(10)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  C  Q
0         0       3    1  22.0      1      0   7.2500        S  0  0
1         1       1    0  38.0      1      0  71.2833        C  1  0
2         1       3    0  26.0      0      0   7.9250        S  0  0
3         1       1    0  35.0      1      0  53.1000        S  0  0
4         0       3    1  35.0      0      0   8.0500        S  0  0
5         0       3    1   NaN      0      0   8.4583        Q  0  1
6         0       1    1  54.0      0      0  51.8625        S  0  0
7         0       3    1   2.0      3      1  21.0750        S  0  0
8         1       3    0  27.0      0      2  11.1333        S  0  0
9         1       2    0  14.0      1      0  30.0708        C  1  0

In [101]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int32  
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  891 non-null    object 
 8   C         891 non-null    uint8  
 9   Q         891 non-null    uint8  
dtypes: float64(2), int32(1), int64(4), object(1), uint8(2)
memory usage: 54.1+ KB


**Column 'Age' remain empty entries. I fill mean of age into these entries**

In [102]:
from math import *

age_mean = floor(np.mean(new_df['Age']))

new_df['Age'].fillna(age_mean, inplace=True)

new_df.head(10)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  C  Q
0         0       3    1  22.0      1      0   7.2500        S  0  0
1         1       1    0  38.0      1      0  71.2833        C  1  0
2         1       3    0  26.0      0      0   7.9250        S  0  0
3         1       1    0  35.0      1      0  53.1000        S  0  0
4         0       3    1  35.0      0      0   8.0500        S  0  0
5         0       3    1  29.0      0      0   8.4583        Q  0  1
6         0       1    1  54.0      0      0  51.8625        S  0  0
7         0       3    1   2.0      3      1  21.0750        S  0  0
8         1       3    0  27.0      0      2  11.1333        S  0  0
9         1       2    0  14.0      1      0  30.0708        C  1  0

In [103]:
new_df.drop('Embarked', axis=1, inplace=True)

new_df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  C  Q
0         0       3    1  22.0      1      0   7.2500  0  0
1         1       1    0  38.0      1      0  71.2833  1  0
2         1       3    0  26.0      0      0   7.9250  0  0
3         1       1    0  35.0      1      0  53.1000  0  0
4         0       3    1  35.0      0      0   8.0500  0  0

**We look the desciption again**

In [104]:
new_df.describe()

Survived      Pclass         Sex         Age       SibSp       Parch  \
count  891.000000  891.000000  891.000000  891.000000  891.000000  891.000000   
mean     0.383838    2.308642    0.647587   29.560236    0.523008    0.381594   
std      0.486592    0.836071    0.477990   13.005010    1.102743    0.806057   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    2.000000    0.000000   22.000000    0.000000    0.000000   
50%      0.000000    3.000000    1.000000   29.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   35.000000    1.000000    0.000000   
max      1.000000    3.000000    1.000000   80.000000    8.000000    6.000000   

             Fare           C           Q  
count  891.000000  891.000000  891.000000  
mean    32.204208    0.188552    0.086420  
std     49.693429    0.391372    0.281141  
min      0.000000    0.000000    0.000000  
25%      7.910400    0.000000    0.000000  
50%     14.454200    0.000000    0.000000  
75%     31.000000    0.000000    0.000000  
max    512.329200    1.000000    1.000000

**No missing value!**

In [105]:
X = new_df.drop('Survived', axis=1)
X.head()

Pclass  Sex   Age  SibSp  Parch     Fare  C  Q
0       3    1  22.0      1      0   7.2500  0  0
1       1    0  38.0      1      0  71.2833  1  0
2       3    0  26.0      0      0   7.9250  0  0
3       1    0  35.0      1      0  53.1000  0  0
4       3    1  35.0      0      0   8.0500  0  0

In [106]:
y = new_df['Survived']
y.head()

0    0
1    1
2    1
3    1
4    0
Name: Survived, dtype: int64

# Build Model

In [107]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

In [68]:
from sklearn.model_selection import cross_val_score

**LogisticRegression**

In [110]:
cross_val_score(LogisticRegression(solver='liblinear',multi_class='ovr'), X, y, cv=5).mean()

0.7923733601154981

**SVC**

In [114]:
cross_val_score(SVC(gamma='auto'), X, y, cv=5).mean()

0.6914757391249765

**RandomForestClassifier**

In [130]:
cross_val_score(RandomForestClassifier(n_estimators=70), X, y, cv=5).mean()

0.8125980792166217

**DecisionTreeClassifier**

In [123]:
cross_val_score(DecisionTreeClassifier(), X, y, cv=5).mean()

0.7789467076768564

## Then we use RandomForestClassifier model

In [131]:
RFC = RandomForestClassifier()

In [132]:
RFC.fit(X, y)

RandomForestClassifier()

**import test file**

In [173]:
testing = pd.read_csv('test_titanic.csv')

testing.head(10)

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   
5          897       3                    Svensson, Mr. Johan Cervin    male   
6          898       3                          Connolly, Miss. Kate  female   
7          899       2                  Caldwell, Mr. Albert Francis    male   
8          900       3     Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female   
9          901       3                       Davies, Mr. John Samuel    male   

    Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
0  34.5      0      0     330911   7.8292   NaN        Q  
1  47.0      1      0     363272   7.0000   NaN        S  
2  62.0      0      0     240276   9.6875   NaN        Q  
3  27.0      0      0     315154   8.6625   NaN        S  
4  22.0      1      1    3101298  12.2875   NaN        S  
5  14.0      0      0       7538   9.2250   NaN        S  
6  30.0      0      0     330972   7.6292   NaN        Q  
7  26.0      1      1     248738  29.0000   NaN        S  
8  18.0      0      0       2657   7.2292   NaN        C  
9  21.0      2      0  A/4 48871  24.1500   NaN        S

In [174]:
testing.drop(['Name', 'Ticket', 'Cabin'], axis=1, inplace=True)

testing.head(10)

PassengerId  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0          892       3    male  34.5      0      0   7.8292        Q
1          893       3  female  47.0      1      0   7.0000        S
2          894       2    male  62.0      0      0   9.6875        Q
3          895       3    male  27.0      0      0   8.6625        S
4          896       3  female  22.0      1      1  12.2875        S
5          897       3    male  14.0      0      0   9.2250        S
6          898       3  female  30.0      0      0   7.6292        Q
7          899       2    male  26.0      1      1  29.0000        S
8          900       3  female  18.0      0      0   7.2292        C
9          901       3    male  21.0      2      0  24.1500        S

In [175]:
testing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Sex          418 non-null    object 
 3   Age          332 non-null    float64
 4   SibSp        418 non-null    int64  
 5   Parch        418 non-null    int64  
 6   Fare         417 non-null    float64
 7   Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(2)
memory usage: 26.2+ KB


In [176]:
testing['Fare'].fillna(testing['Fare'].mean(), inplace=True)

In [177]:
testing['Sex'] = LE.fit_transform(testing['Sex'])

testing.head(10)

PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0          892       3    1  34.5      0      0   7.8292        Q
1          893       3    0  47.0      1      0   7.0000        S
2          894       2    1  62.0      0      0   9.6875        Q
3          895       3    1  27.0      0      0   8.6625        S
4          896       3    0  22.0      1      1  12.2875        S
5          897       3    1  14.0      0      0   9.2250        S
6          898       3    0  30.0      0      0   7.6292        Q
7          899       2    1  26.0      1      1  29.0000        S
8          900       3    0  18.0      0      0   7.2292        C
9          901       3    1  21.0      2      0  24.1500        S

In [178]:
testing['Age'].fillna(floor(testing['Age'].mean()), inplace=True)

testing.head()

PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0          892       3    1  34.5      0      0   7.8292        Q
1          893       3    0  47.0      1      0   7.0000        S
2          894       2    1  62.0      0      0   9.6875        Q
3          895       3    1  27.0      0      0   8.6625        S
4          896       3    0  22.0      1      1  12.2875        S

In [179]:
dum = pd.get_dummies(df['Embarked'])

dum.head(5)

C  Q  S
0  0  0  1
1  1  0  0
2  0  0  1
3  0  0  1
4  0  0  1

In [180]:
dum.drop('S', axis=1, inplace=True)
dum.head()

C  Q
0  0  0
1  1  0
2  0  0
3  0  0
4  0  0

In [181]:
testing['C'] = dum['C']
testing['Q'] = dum['Q']

testing.head(10)

PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  C  Q
0          892       3    1  34.5      0      0   7.8292        Q  0  0
1          893       3    0  47.0      1      0   7.0000        S  1  0
2          894       2    1  62.0      0      0   9.6875        Q  0  0
3          895       3    1  27.0      0      0   8.6625        S  0  0
4          896       3    0  22.0      1      1  12.2875        S  0  0
5          897       3    1  14.0      0      0   9.2250        S  0  1
6          898       3    0  30.0      0      0   7.6292        Q  0  0
7          899       2    1  26.0      1      1  29.0000        S  0  0
8          900       3    0  18.0      0      0   7.2292        C  0  0
9          901       3    1  21.0      2      0  24.1500        S  1  0

In [182]:
X = testing.drop(['PassengerId', 'Embarked'], axis=1)
X.head()

Pclass  Sex   Age  SibSp  Parch     Fare  C  Q
0       3    1  34.5      0      0   7.8292  0  0
1       3    0  47.0      1      0   7.0000  1  0
2       2    1  62.0      0      0   9.6875  0  0
3       3    1  27.0      0      0   8.6625  0  0
4       3    0  22.0      1      1  12.2875  0  0

In [183]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Pclass  418 non-null    int64  
 1   Sex     418 non-null    int32  
 2   Age     418 non-null    float64
 3   SibSp   418 non-null    int64  
 4   Parch   418 non-null    int64  
 5   Fare    418 non-null    float64
 6   C       418 non-null    uint8  
 7   Q       418 non-null    uint8  
dtypes: float64(2), int32(1), int64(3), uint8(2)
memory usage: 18.9 KB


In [184]:
y_predict = RFC.predict(X)

y_predict

array([0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [190]:
result = pd.DataFrame({"PassengerId": testing['PassengerId'], "Survived": y_predict})

result.head(10)

PassengerId  Survived
0          892         0
1          893         0
2          894         1
3          895         1
4          896         0
5          897         0
6          898         0
7          899         0
8          900         1
9          901         0

In [192]:
result.to_csv("submission.csv", index=False)